### PR TITLE
Block Tap/Touch Gesture when hint/linked clicked

### LINKED ISSUE #6141: [Feature Request] - Allow clicking on links when tap gestures are enabled
Requested at: https://groups.google.com/forum/#!topic/anki-android/3IhZeImruSs

Currently it is not possible to click on hints if tap gestures are enabled. Instead only the taps are registered and the answer is shown.

Would it be possible to add a greater degree of accuracy so that hints could be clicked but taps on other parts of the screen would cause the answer to be shown?

### PR DESCRIPTION
**Pull request by david-allison-1**
## Disclosure

I am accepting a financial contribution for this feature request

## Purpose / Description
A user has requested that link clicks do not trigger the "touch" gestures.

## Fixes
Fixes #6141 

## Approach

Complex : we can get a `HitTest()` from the browser, but only inside a touch handler.

So, add the touch handler, filter to only the events that we want, and stop gesture processing on events where the `hitTest` was a link or an image with src.

## How Has This Been Tested?

Tested on my phone. Also tested with option off to remove regressions

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 1]
```diff
@@ -2635,8 +2654,105 @@ public boolean onSingleTapConfirmed(MotionEvent e) {
             showLookupButtonIfNeeded();
             return false;
         }
+
+        public void onWebViewCreated(@NonNull WebView webView) {
+
+        }
+
+        public void onFillFlashcard() {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:2662]: Codacy Issue found: [Document empty method body](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5538529)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5538529&bid=18155835)

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 2]
```diff
@@ -2635,8 +2654,105 @@ public boolean onSingleTapConfirmed(MotionEvent e) {
             showLookupButtonIfNeeded();
             return false;
         }
+
+        public void onWebViewCreated(@NonNull WebView webView) {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:2658]: Codacy Issue found: [Document empty method body](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5538529)

File: AnkiDroid/src/main/res/values/10-preferences.xml [hunk 3]
```diff
@@ -73,6 +73,8 @@
     </string-array>
     <string name="gestures">Enable gestures</string>
     <string name="gestures_summ">Assign gestures to actions such as answering and editing cards.</string>
+    <string name="gestures_allow_links">Ignore touch gestures on links</string>
+    <string name="gestures_allow_links_summary">Touch gestures will not be processed if a link/hint is pressed.</string>
```
Comment (mikehardy) [inline AnkiDroid/src/main/res/values/10-preferences.xml:77]: Is it worth noting that it may slow the gesture system down a lot? (any idea how much?)

**Comment by mikehardy** [comment]
This looks fine to me, I'm prepare to accept it - I just had one question about exposing the performance tradeoff to users, and the main reason I'm writing here now is to think out loud on an idea I've had - is it possible on app startup, for people with analytics enabled, to send a quick summary of "interesting configuration elements" - which I interpret as just config items we're curious if people are using or not. If it is (it should be?) then it would be interesting to tag this (as well as e.g. custom language, notifications, v2 scheduler, maybe others?) just to see what's getting used and what isn't. That could help justify future maintenance if it becomes a burden (shouldn't but you never know)

**Comment by david-allison-1** [comment]
Added in 6207, will get perf now.

**Comment by mikehardy** [comment]
sounds good - thanks!

**Comment by david-allison-1** [comment]
Well... we're definitely not gonna do HFT with it... but it's probably not worth telling the user about (or even commenting).

 | Before | After|
 |     :---      |          :--- |
 | 3ms     | 6ms    |
 | 4ms       | 12ms     |
 | 3ms       | 10ms      |

I'll remove the comments. Great catch 💯

**Comment by david-allison-1** [comment]
Regarding Lint checks (all opinions my own):

* Ideal would be to get reporting inside the PR - not sure how easy this is

----

https://travis-ci.org/github/ankidroid/Anki-Android/jobs/688181348

`[SyntheticAccessor]` I would propose disabling the rule completely, or at least moving it to a warning - synthetic methods - fine, make them `package`. Synthetic variables? I don't think we care about the perf hit (if any), and the only issue will be hitting the Dalvik 64k limit, which is annoying, but will happen regardless.

```
/home/travis/build/ankidroid/Anki-Android/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:444: Error: Access to private field mGestureDetectorImpl of class AbstractFlashcardViewer requires synthetic accessor [SyntheticAccessor]
            if (!mGestureDetectorImpl.eventCanBeSentToWebView(event)) {
                 ~~~~~~~~~~~~~~~~~~~~
```

`[MissingTranslation]` - This x2 - not to be handled, it's fine
```
/home/travis/build/ankidroid/Anki-Android/AnkiDroid/src/main/res/values/10-preferences.xml:77: Error: "gestures_allow_links_summary" is not translated in "nn" (Norwegian Nynorsk), "tt" (Tatar), "de" (German), "hi" (Hindi), "no" (Norwegian), "fi" (Finnish), "ru" (Russian), "pt" (Portuguese), "bg" (Bulgarian), "lt" (Lithuanian), "lv" (Latvian), "fr" (French), "hu" (Hungarian), "uk" (Ukrainian), "sk" (Slovak), "sl" (Slovenian), "id" (Indonesian), "ca" (Catalan), "sr" (Serbian), "sv" (Swedish), "ko" (Korean), "gl" (Galician), "el" (Greek), "go", "eo" (Esperanto), "it" (Italian), "es" (Spanish), "zh" (Chinese), "et" (Estonian), "cs" (Czech), "ar" (Arabic), "vi" (Vietnamese), "th" (Thai), "ja" (Japanese), "fa" (Persian), "pl" (Polish), "he" (Hebrew), "ro" (Romanian), "nl" (Dutch), "tr" (Turkish) [MissingTranslation]
    <string name="gestures_allow_links_summary">Touch gestures will not be processed if a link/hint is pressed.</string>
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

`[NewerVersionAvailable]` Not my problem to worry about, maybe if the user adds it in with a newer version, maybe as a TODO to only do it if `build.gradle` is updated
```
/home/travis/build/ankidroid/Anki-Android/AnkiDroid/build.gradle:153: Error: A newer version of com.squareup.okhttp3:okhttp than 3.12.11 is available: 4.7.0 [NewerVersionAvailable]
    implementation 'com.squareup.okhttp3:okhttp:3.12.11'
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by mikehardy** [comment]
Yeah, this will be a tuning process for sure.

- I don't care about synthetic accessors, feel free to PR a change to lint rules to ignore that/those
- I like the missing translation one, it reminds me when merging the PR to sync translations and will serve as a continuous reminder to do so, that's important
- that specific one was a mistake on my part, I had a no inspection on that and removed it. We must not move from the older version until the APIs mentioned in the related comment next to it, feel free to PR an inspection disable for that line

**Comment by david-allison-1** [comment]
Good to go from my end, will get the lint fixes at a later time

**Comment by mikehardy** [comment]
fair enough. I picture that lint-debug one being a work queue basically, growing over time and periodically squashed. Squash any time

API28 CI failure was an emulator that did not fully start. Not sure what exactly to do about that but it may (over time) turn out to the current tall poppy to chop. If so that would be some sort of timeout and re-try I suppose. I just manually kicked it this time.

**Comment by mikehardy** [comment]
released in 2.11alpha5 / 2.10.3beta1 (I picked it so it will be generally available, but it's still beta right now - please test)

**Pull request merged**